core/store/orm/config.go

```diff
@@ -1163,9 +1164,37 @@ func (c Config) P2PListenIP() net.IP {
 	return c.getWithFallback("P2PListenIP", parseIP).(net.IP)
 }
 
-// P2PListenPort is the port that libp2p willl bind to and listen on
-func (c Config) P2PListenPort() uint16 {
-	return uint16(c.viper.GetUint32(EnvVarName("P2PListenPort")))
+// P2PListenPort is the port that libp2p will bind to and listen on
+func (c *Config) P2PListenPort() uint16 {
+	if c.viper.IsSet(EnvVarName("P2PListenPort")) {
+		return uint16(c.viper.GetUint32(EnvVarName("P2PListenPort")))
+	}
+	// Fast path in case it was already set
+	c.randomP2PPortMtx.RLock()
+	if c.randomP2PPort > 0 {
+		c.randomP2PPortMtx.RUnlock()
+		return c.randomP2PPort
+	}
+	c.randomP2PPortMtx.RUnlock()
```

Comment:
hm any reason for not just moving these up, instead of locking / unlocking twice?
```
	c.randomP2PPortMtx.Lock()
	defer c.randomP2PPortMtx.Unlock()
```

Reply:
Yes, I was hoping the comment would explain. It's marginally faster in the most common case that it was already set, since RLock allows concurrent reads.

Reply:
ah, i missed it's RLock vs Lock, cool